marconi/src/Marconi/Index/AddressDatum.hs

```diff
@@ -0,0 +1,517 @@
+-- add an entry in the `datumhash_datum` table):
+--
+--     * a different transaction output has an inline datum with that same hash
+--
+--     * a datum with that same hash has been found in the in transaction body
+--
+--     * a datum with that same hash was included in the witnesses for a Plutus spending script
+--     which was included in the transaction body
+--
+module Marconi.Index.AddressDatum
+  ( -- * AddressDatumIndex
+    AddressDatumIndex
+  , AddressDatumHandle
+  , StorableEvent(..)
+  , StorableQuery(..)
+  , StorableResult(..)
+  , toAddressDatumIndexEvent
+  , AddressDatumQuery
+  , AddressDatumResult
+  , AddressDatumDepth (..)
+  , open
+  ) where
+
+import Cardano.Api qualified as C
+import Cardano.Api.Shelley qualified as C
+import Cardano.Ledger.Alonzo.TxWitness qualified as Ledger
+import Control.Applicative ((<|>))
+import Control.Monad (forM, forM_)
+import Data.Foldable (Foldable (foldl'), fold, toList)
+import Data.List.NonEmpty (NonEmpty)
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Maybe (catMaybes, listToMaybe, mapMaybe)
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Database.SQLite.Simple qualified as SQL
+import Database.SQLite.Simple.ToField qualified as SQL
+import GHC.Generics (Generic)
+import Marconi.Orphans ()
+import RewindableIndex.Storable (Buffered (persistToStorage), HasPoint (getPoint),
+                                 QueryInterval (QEverything, QInterval), Queryable (queryStorage), Resumable,
+                                 Rewindable (rewindStorage), StorableEvent, StorableMonad, StorablePoint, StorableQuery,
+                                 StorableResult, emptyState, filterWithQueryInterval)
+import RewindableIndex.Storable qualified as Storable
+import Text.RawString.QQ (r)
+
+-- | Define the `handler` data type, meant as a wrapper for the connection type (in this case the
+-- SQLite connection). In this indexer, we also add the number of events that we want to return from
+-- the on-disk buffer.
+data AddressDatumHandle = AddressDatumHandle
+    { addressDatumHandleConnection :: SQL.Connection
+    , _addressDatumHandleDiskStore :: Int
+    }
+
+type instance StorableMonad AddressDatumHandle = IO
+
+-- | 'StorableEvent AddressDatumHandle is the type of events. Events are the data atoms that the
+-- indexer consumes.
+-- They depend on the `handle` because they need to eventually be persisted in the database, so the
+-- database has to be able to accomodate them.
+
+-- We store the datum hashes of each address that we processed.
+-- Then we keep a separate 'Map' which stores the actual datum given a datum hash.
+-- Note that we don't always have the actual datum for a given hash.
+data instance StorableEvent AddressDatumHandle =
+    AddressDatumIndexEvent
+        (Map C.AddressAny (Set (C.Hash C.ScriptData)))
+        (Map (C.Hash C.ScriptData) C.ScriptData)
+        !C.ChainPoint
```

Comment:
newbie question :) 
I am not clear as to why we need strictness here?

Reply:
I always put strictness by default (that was a practice I read a while ago)